### PR TITLE
feat: added check for wait duration properties

### PR DESCRIPTION
**Pull request by massfords**
refactor: leverage common code between exclusive prop checks

### CODE REVIEW

File: src/__tests__/definitions/invalid-wait-duration.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    "wait_using_seconds": {
+      "Type": "Wait",
+      "Seconds": 10,
+      "SecondsPath": "$.expiryseconds",
+      "Next": "wait_using_timestamp"
+    },
+    "wait_using_timestamp": {
+      "Type": "Wait",
+      "Timestamp": "2015-09-04T01:59:00Z",
+      "TimestampPath": "$.expirydate",
```
Comment (massfords) [inline src/__tests__/definitions/invalid-wait-duration.json:19]: previously, this wasn't being reported

File: src/checks/exactlyOneOfProperty.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import {State, StateMachine, StateMachineError, StateMachineErrorCode} from "../types";
+import {getStates, getStatesContainer, StateEntry} from "./get-states";
+
+export const exactlyOneOfPropertyWorkflow = ({definition, filter, props, errorCode}: {
+    definition: StateMachine,
+    filter: (entry: StateEntry) => boolean
+    props: string[],
+    errorCode: StateMachineErrorCode
+}): StateMachineError[] => {
```
Comment (massfords) [inline src/checks/exactlyOneOfProperty.ts:9]: refactored the existing terminal check into a common helper function to re-use for other mutual exclusive property checks

using this now with Next/End and WaitDuration

File: src/checks/exactlyOneOfProperty.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        .reduce((prev, curr) => prev + curr, 0)
+    if (count !== 1) {
+        return {
+            "Error code": errorCode,
+            // Use of JSONPath within the error message is unnecessary
+            // since the state names are unique.
+            Message: `State "${stateName}" MUST contain exactly one of ${props.map((p) => {
+                return `"${p}"`
+            }).join(", ")}`,
+
```
Comment (massfords) [inline src/checks/exactlyOneOfProperty.ts:46]: some sample output when modifying the tests to produce an error:

```
+ Array [
+   Object {
+     "Error code": "WAIT_DURATION",
+     "Message": "State \"wait_using_timestamp\" MUST contain exactly one of \"Seconds\", \"SecondsPath\", \"Timestamp\", \"TimestampPath\"",
+   },
+ ]
```

```
+ Array [
+   Object {
+     "Error code": "TERMINAL_STATE_WITH_NEXT",
+     "Message": "State \"FinalState\" MUST contain exactly one of \"End\", \"Next\"",
+   },
+ ]
```

File: src/checks/terminalStateWithNext.ts [hunk 4]
```diff
@@ -9,22 +9,14 @@ import {getStates, getStatesContainer} from "./get-states";
...
-        })
+    return exactlyOneOfPropertyWorkflow({
+        props: ['End', 'Next'],
+        errorCode: StateMachineErrorCode.TerminalStateWithNextError,
+        definition,
+        filter: ({state}) => {
+            return state.Type !== 'Succeed' &&
+                state.Type !== 'Fail' &&
+                state.Type !== 'Choice'
+        }
```
Comment (massfords) [inline src/checks/terminalStateWithNext.ts:20]: use the workflow helper function to do the check

**Comment by massfords** [comment]
@ChristopheBougere build fails w/ a commit message error. I had missed the merge of my other PR so perhaps it doesn't like the merge. Feel free to squash

I can look at the other issues I added later this week. I think the refactoring in this PR to have a common function for doing the exclusive prop check will make the other issues easier to close.

**Review by ChristopheBougere — APPROVED** [review]
Much cleaner, thanks!

**Pull request merged**